tools/clang/tools/dxcompiler/dxcompilerobj.cpp

```diff
@@ -476,48 +450,107 @@ class HLSLExtensionsCodegenHelperImpl : public HLSLExtensionsCodegenHelper {
   }
 
 public:
-  HLSLExtensionsCodegenHelperImpl(CompilerInstance &CI, DxcLangExtensionsHelper &langExtensionsHelper, StringRef rootSigDefine)
+  HLSLExtensionsCodegenHelperImpl(CompilerInstance &CI, DxcLangExtensionsHelper &langExtensionsHelper, StringRef rootSigDefine, const hlsl::options::OptimizationToggles &hlslOptToggles, bool HLSLEnableLifetimeMarkers)
   : m_CI(CI), m_langExtensionsHelper(langExtensionsHelper)
   , m_rootSigDefine(rootSigDefine)
+  , m_HLSLOptToggles(hlslOptToggles)
+  , m_HLSLEnableLifetimeMarkers(HLSLEnableLifetimeMarkers)
   {}
 
   // Write semantic defines as metadata in the module.
   virtual void WriteSemanticDefines(llvm::Module *M) override {
+    WriteSemanticDefines(M, m_semanticDefines);
+  }
+  // Update CodeGenOption based on HLSLOptimizationToggles.
+  void UpdateSemanticDefinesAndOptToggles() override {
+    // This shouldn't really be called more than once, but just in case it is,
```

Comment:
Should we add an assert here to catch cases where it is accidentally getting called twice?